pkg/registry/pod/rest.go

```diff
@@ -81,9 +81,28 @@ func (podStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fiel
 	return append(errorList, validation.ValidatePodUpdate(obj.(*api.Pod), old.(*api.Pod))...)
 }
 
-// CheckGracefulDelete allows a pod to be gracefully deleted.
+// CheckGracefulDelete allows a pod to be gracefully deleted. It updates the DeleteOptions to
+// reflect the desired grace value.
 func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
-	return false
+	if options == nil {
+		return false
+	}
+	pod := obj.(*api.Pod)
+	// use the user provided value
+	if options.GracePeriodSeconds != nil {
+		period := int64(*options.GracePeriodSeconds)
+		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
+		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
+			pod.Spec.TerminationGracePeriodSeconds = &period
```

Comment:
> On May 28, 2015, at 1:38 AM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/registry/pod/rest.go:
> 
> >  func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
> > -	return false
> > +	if options == nil {
> > +		return false
> > +	}
> > +	pod := obj.(*api.Pod)
> > +	// use the user provided value
> > +	if options.GracePeriodSeconds != nil {
> > +		period := int64(*options.GracePeriodSeconds)
> > +		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
> > +		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
> > +			pod.Spec.TerminationGracePeriodSeconds = &period
> Ok, I can see the point that the DeleteOptions are modifying the desired state, but it definitely adds a wrinkle to configuration reconciliation #1702.
> 
> This may not be an issue for controller-generated pods, since they'll manage terminationGracePeriodSeconds in the pod template rather than in the pods themselves, but I want to get the principle sorted out.
> 
> This isn't the only known instance where modifying an initialized field will happen -- e.g., auto-scaling will modify an initial replicas count and auto-sizing will modify initial cpu and memory values -- but either those values would be defaults from the beginning, or initial value "hints" could be provided via an annotation or in the auto-scaling/sizing configs. In either case, the config mechanism would know that the user didn't intend to configure those fields, since the user's config would not contain the initial values.
> 
> If a user did have a pod managed by a declarative config file, they might legitimately want to declaratively update terminationGracePeriodSeconds, until it is set imperatively via DeleteOptions. I can tell you from experience that there will be scenarios where someone wants to update a pod during a (sufficiently long) grace period.
> 
> The (declarative) semantics that I'm accustomed to, and that proposed in #1702, #1007, and elsewhere, is that removal of a field from a config implies that it should be restored to its default value before dropping it from the set of declaratively managed fields. If it were acceptable to require users to explicitly specify the updated value they expected or to explicitly nullify the field, as in a patch, in at least one declarative update, then that would make patch generation significantly simpler.
> 
So here's another thought.  This field is special in that it should only be shortened once delete timestamp is set.  That behavior needs to be documented in the schema - may not be increased post deletion - and config reconciliation probably has to take that into account.  If a resource has already been marked deleted, should we be limiting certain fields from being updated?  Right now we allow all changes that update allows, but does that potentially impact the expectation users set?  

Alternatively, if instead we ignore updates to the fields on the apiserver post deletion, it does not matter what config reconciliation sends.  Ultimately that logic has to exist on a certain field (whether it's spec.TerminationPeriodInSeconds or metadata.DeleteInSeconds) if we want to enforce the "only shrink grace period post deletion" condition.  Is it more or less confusing to API clients to ignore a change to that field in certain cases?  It seems like a net reduction in overall complexity to ignore certain changes for clients, although it does reduce predictability.
> We could potentially treat omission of a field simply as dropping it from the set of declaratively mana ged fiel ds, though I have no idea whether users would find that unintuitive. We'd still need to detect declarative removal of map and list elements, IMO, such as removal of a label or of a container or volume, so we couldn't just use the annotation containing the config to determine that the configuration changed, as I suggested for Kubelet's mirror pods (#8683).
> 
> One challenge with using a magic map key as a json/yaml annotation in user configs to indicate that certain fields shouldn't be diff'ed during patch generation (e.g., $noupdate) is that we can only use the same key once per map. I suppose it could specify a list of field names.
> 
> cc @ghodss for his thoughts
> 
> —
> Reply to this email directly or view it on GitHub.
>